### ISSUE TITLE
Build an image with the database already configured.

### ISSUE THREAD

**Issue by stewartbryson** [oracle/docker-images#1474]
It seems silly to say this... but a docker image that requires 10 minutes to start is just plain ridicules. And unusable. Images of MySQL, Aurora, Postgres, etc. all start and are available in seconds.

At the heart of this is the fact that you think it's ok to create/configure the database during "docker run" instead of "docker build". The database should be created and configured during "docker build". I don't understand why you believe this is not an option.

Just use dynamic listener registration instead of a listener.ora file, and the dynamic docker hostname should be detected and used at startup time.

**Comment by ggajos** [comment]
Yes it is. Not sure if it's suitable for you but you can find here instruction how to speed this up: https://medium.com/@ggajos/drop-db-startup-time-from-45-to-3-minutes-in-dockerized-oracle-19-3-0-552068593deb

**Comment by gvenzl** [comment]
@stewartbryson, it's not a matter of dynamic listener configuration or listener.ora, it's to do with configuration and persistence (and other people complaining about image size).

The database character set has to be specified on DB configuration and cannot be changed anymore later on. While AL32UTF8 may be good enough for you, there are plenty of other users out there that have said AL32UTF8 is not an option for them. We didn't even have the option to specify the character set in the beginning and got a lot of feedback that it is needed.
The same goes for the SID. We didn't think we need to provide an option to specify your SID, it's a container, what do you care, right? Well, think again. Many folks had the need or wanted to have their SIDs the way they specified and so we also gave that option.

Docker isn't smart enough to copy files inside a volume to the outside world when using host-mounted volumes, instead of names volumes. If you do so, still all your data files disappear and your database is gone, even if you had it in the image to begin with. That's the very reason why you have to update the Dockerfile to remove the `VOLUME` instruction because guess what, you can't do that in an inherited image either (nor the `EXPOSE` instruction).

Now, for a long time have we provided a sample on how you can build an image with the DB inside, see https://github.com/oracle/docker-images/tree/master/OracleDatabase/SingleInstance/samples/prebuiltdb
You need to modify the Dockerfile once and commit the image once, is that really too much effort?

**Comment by Lordroran** [comment]
Is it possible to automate this within a docker build? https://medium.com/@ggajos/drop-db-startup-time-from-45-to-3-minutes-in-dockerized-oracle-19-3-0-552068593deb

**Comment by stewartbryson** [comment]
> The database character set has to be specified on DB configuration and cannot be changed anymore later on. While AL32UTF8 may be good enough for you, there are plenty of other users out there that have said AL32UTF8 is not an option for them. We didn't even have the option to specify the character set in the beginning and got a lot of feedback that it is needed.
> The same goes for the SID. We didn't think we need to provide an option to specify your SID, it's a container, what do you care, right? Well, think again. Many folks had the need or wanted to have their SIDs the way they specified and so we also gave that option.

I asked you to "create an image"... not "stop creating anything else and only create this one image". By all means, continue to produce what you currently produce... many people with customization requirements will find these images useful. But start providing a few already configured images with a few standard charactersets and a constant SID name.

> Now, for a long time have we provided a sample on how you can build an image with the DB inside, see https://github.com/oracle/docker-images/tree/master/OracleDatabase/SingleInstance/samples/prebuiltdb
> You need to modify the Dockerfile once and commit the image once, is that really too much effort?

All of your competitors (you know... the ones with increasing market share) provide this, and you are the only ones that don't. So I would ask you the same question... is this really too much effort for you to produce?

**Comment by Djelibeybi** [comment]
@stewartbryson I understand that you might be somewhat frustrated, but I don't think passive-aggressive comparisons to competitors is warranted. Especially considering Oracle has been publicly publishing our Dockerfiles for significantly longer than any of them and has certified our products in containers for longer again.

Let's just stick to the topic at hand, shall we? And on that note:  we welcome and accept submissions to this repo, so pull requests (that meet the contribution guide) welcome!

**Comment by stewartbryson** [comment]
> @stewartbryson I understand that you might be somewhat frustrated, but I don't think passive-aggressive comparisons to competitors is warranted. Especially considering Oracle has been publicly publishing our Dockerfiles for significantly longer than any of them and has certified our products in containers for longer again.

Just let me know when I can `docker run oracle`... because I can do that with every other... Oh yeah, sorry for comparing you again.

**Comment by Lordroran** [comment]
Shall i open a new Issue for my Question then?

**Comment by stewartbryson** [comment]
> Shall i open a new Issue for my Question then?

I'm happy to reopen @Lordroran if my Issue description is in-line with your request. I think the maintainers are hoping for a PR from someone and probably aren't going to prioritize this. I've moved all my ODI-based images to use MySQL instead of Oracle... but I can't do that with other FMW products. So I still use Google compute instances in those areas where the spin-up time is too excruciating to use these images.

So... shall I reopen?

**Comment by Lordroran** [comment]
I would like to have this Issue reopened, because i do not see a possibility to automate this with the current images.

**Comment by Djelibeybi** [comment]
@gvenzl and I discussed this and I believe he is working on updating the Dockerfiles to remove the `EXPOSE` and `VOLUME` directives that are no longer necessary. This will allow for automatic Database creation during build to persist to the image. 

Though, I suspect that changing the `ORACLE_BASE` and `ORACLE_HOME` parameters from `/opt/oracle` would probably do the same thing right now, i.e. just move the `oradata` directory out from the `VOLUME` location.

**Comment by Lordroran** [comment]
> Though, I suspect that changing the `ORACLE_BASE` and `ORACLE_HOME` parameters from `/opt/oracle` would probably do the same thing right now, i.e. just move the `oradata` directory out from the `VOLUME` location.

I suspected this, but i have troubles starting Oracle while at build time. Last time i tried the script did not return and i was stuck in the start script. 

Do you have an example for starting Oracle manually?

**Comment by Djelibeybi** [comment]
No, but if changing the environment variables causes the container not to start correctly, I would consider that a bug.

**Comment by Lordroran** [comment]
This is not about changing environment variables. This is about starting the DB while doing a dockerbuild. We need to start Oracle during the dockerbuild to initialize the DB. 
 
I tried to do this by hand with either /opt/oracle/runOracle.sh or /opt/oracle/createDB.sh, but both of the scripts do not seem to terminate after the DB is ready.

**Comment by stewartbryson** [comment]
> This is not about changing environment variables. This is about starting the DB while doing a dockerbuild. We need to start Oracle during the dockerbuild to initialize the DB.
> 
> I tried to do this by hand with either /opt/oracle/runOracle.sh or /opt/oracle/createDB.sh, but both of the scripts do not seem to terminate after the DB is ready.

`runOracle.sh` is designed to run perpetually... it's what is specified as the `CMD` in the Dockerfile.
You can simply remove the `tail -f` part from this script so that it completes, and then use`runOracle.sh && tail -f <whatever>` as the CMD and it should be equivalent.

**Comment by gggeek** [comment]
My 2c: even if it results in a big image, I'd warmly welcome containers with a pre-built DB that I can run with a simple `docker run oraclexxx`.
No need for custom service names or character sets.

Atm I am dabbling with adding oracle to a docker-compose stack, and the download+build time is quite long (I have basically forked the oracle 18 container inside my project).

For persisting data, I use host mounts instead of volumes, and I think that it is not hard, for anyone who desires so, to add to her own docker shell scripts or docker-compose config the bit that brings back the data files to the host. The main problem seems to be with how the db is not natively capable of (re)creating the datafiles if those are missing at boot.
This could be fixable f.e. by keeping copies of empty data files in a separate location and having a shell script running at container start that copies them over if oradata dir is empty. Kinda like what runOracle.sh currently does (except for the fact that runOracle.sh moves the existing files, which makes it work only once and break if the volume gets manually cleaned up before a container restart).
Maybe theer is even a better, native solution that I am not aware of that would cretae data files... my oracle-fu is getting rusty

**Comment by Lordroran** [comment]
> `runOracle.sh` is designed to run perpetually... it's what is specified as the `CMD` in the Dockerfile.
> You can simply remove the `tail -f` part from this script so that it completes, and then use`runOracle.sh && tail -f <whatever>` as the CMD and it should be equivalent.

My Problem right now is that even if i remove the tail part, the script calls the other script createDB.sh and this script doesnt return. I dont know if i did not wait enough, but "docker ps" says its over one hour. The last output message is "Completing Database Creation".

**Comment by gggeek** [comment]
@Lordroran I have done presumably the exact same thing: in my container, runOracle.sh has been customized with a simple removal of the tail/wait part. the container's ENTRYPOINT and CMD are changed to run my own shell script, which does call runOracle.sh and then goes waiting on its own.
It does work, but it takes a long time on the first start of the container. I am using a 16GB, 2 CPU cores laptop, running docker in a linux VM on windows, and I get 5-10 minutes for db creation.

The code can be seen here: https://github.com/gggeek/db-3v4l/tree/oracle_support/docker/images/oracle/18.4

And here is the output of 'docker log'

```
oracle_18_4_1            | [Thu Mar  5 13:57:40 UTC 2020] Handing over control to runOracle.sh...
oracle_18_4_1            | ORACLE PASSWORD FOR SYS AND SYSTEM: 3v4l
oracle_18_4_1            | Specify a password to be used for database accounts. Oracle recommends that the password entered should be at least 8 characters in length, contain at least 1 uppercase character, 1 lower case character and 1 digit [0-9]. Note that the same password will be used for SYS, SYSTEM and PDBADMIN accounts:
oracle_18_4_1            | Confirm the password:
oracle_18_4_1            | Configuring Oracle Listener.
oracle_18_4_1            | Listener configuration succeeded.
oracle_18_4_1            | Configuring Oracle Database XE.
oracle_18_4_1            | Enter SYS user password:
oracle_18_4_1            |******
oracle_18_4_1            | Enter SYSTEM user password:
oracle_18_4_1            | ***
oracle_18_4_1            | Enter PDBADMIN User Password:
oracle_18_4_1            | ***
oracle_18_4_1            | Prepare for db operation
oracle_18_4_1            | 7% complete
oracle_18_4_1            | Copying database files
oracle_18_4_1            | 29% complete
oracle_18_4_1            | Creating and starting Oracle instance
oracle_18_4_1            | 30% complete
oracle_18_4_1            | 31% complete
oracle_18_4_1            | 34% complete
oracle_18_4_1            | 38% complete
oracle_18_4_1            | 41% complete
oracle_18_4_1            | 43% complete
oracle_18_4_1            | Completing Database Creation
oracle_18_4_1            | 47% complete
oracle_18_4_1            | 50% complete
oracle_18_4_1            | Creating Pluggable Databases
oracle_18_4_1            | 54% complete
oracle_18_4_1            | 71% complete
oracle_18_4_1            | Executing Post Configuration Actions
oracle_18_4_1            | 93% complete
oracle_18_4_1            | Running Custom Scripts
oracle_18_4_1            | 100% complete
oracle_18_4_1            | Database creation complete. For details check the logfiles at:
oracle_18_4_1            |  /opt/oracle/cfgtoollogs/dbca/XE.
oracle_18_4_1            | Database Information:
oracle_18_4_1            | Global Database Name:XE
oracle_18_4_1            | System Identifier(SID):XE
oracle_18_4_1            | Look at the log file "/opt/oracle/cfgtoollogs/dbca/XE/XE.log" for further details.
oracle_18_4_1            |
oracle_18_4_1            | Connect to Oracle Database using one of the connect strings:
oracle_18_4_1            |      Pluggable database: oracle_18_4/XEPDB1
oracle_18_4_1            |      Multitenant container database: oracle_18_4
oracle_18_4_1            | Use https://localhost:5500/em to access Oracle Enterprise Manager for Oracle Database XE
oracle_18_4_1            | The Oracle base remains unchanged with value /opt/oracle
oracle_18_4_1            | #########################
oracle_18_4_1            | DATABASE IS READY TO USE!
oracle_18_4_1            | #########################
```

**Comment by gggeek** [comment]
Btw: I did not test this theory yet, but it might be that merging this PR ( https://github.com/oracle/docker-images/pull/1537 ) could make it easier to share a pre-built container which has a ready-to-use database:

- the original container gets built with a ready-to-run-db, having a predetermined sys password and charset. The image is big in size, but downloading it once is enough to run it many times locally. Data is only persisted inside the container, so it is not recommended to be used as-is for persistent scenarios. Resetting the container to its pristine image corresponds to a db reset.
- once the end user has downloaded it locally, she can add opt for adding a docker mount/volume that maps to '/oradata' inside the container. When the container is started the next time, the db will be freshly recreated (without need to reinstall the oracle software), and from then on it will be persisted on the host

**Comment by Lordroran** [comment]
> 
> 
> > `runOracle.sh` is designed to run perpetually... it's what is specified as the `CMD` in the Dockerfile.
> > You can simply remove the `tail -f` part from this script so that it completes, and then use`runOracle.sh && tail -f <whatever>` as the CMD and it should be equivalent.
> 
> My Problem right now is that even if i remove the tail part, the script calls the other script createDB.sh and this script doesnt return. I dont know if i did not wait enough, but "docker ps" says its over one hour. The last output message is "Completing Database Creation".

Nevermind... I just came back from lunch and it says 51% complete. So it is just still running.

**Comment by troopa81** [comment]
I need to build an already configured Oracle image to speed up Oracle database start-up during CI tests.

I let this here if anyone one wants to build an already configured Oracle image, please let me know if you know a better way to do.

You have to first checkout this [branch](https://github.com/troopa81/docker-images/tree/already_configured_partial) or apply the following [patch](https://github.com/troopa81/docker-images/commit/8552b6f20c4cb53947b2d057762d0fd556079854)

Then type the following

```shell
$ ./buildDockerImage.sh -v 18.4.0 -x
$ docker run -d  -e ORACLE_SID="XE" -e ORACLE_PWD="adminpass" -e ORACLE_PDB="ORCLPDBTEST" -e ORACLE_CHARACTERSET="AL32UTF8" -p 0.0.0.0:1521:1521 oracle/database:18.4.0-xe
96764ef54ac7e4d6144f5e13d01f1dd8535d4ef743c07c82ebfca16aaea795bc
$ # Wait for Oracle to be completely started
$ docker logs 96764ef54ac7e4d6144f5e13d01f1dd8535d4ef743c07c82ebfca16aaea795bc
$ docker stop 96764ef54ac7e4d6144f5e13d01f1dd8535d4ef743c07c82ebfca16aaea795bc 
$ docker commit 96764ef54ac7e4d6144f5e13d01f1dd8535d4ef743c07c82ebfca16aaea795bc oracle-configured
```

You could have changed the ORACLE_BASE environnement variable from the `docker build` command line, but you 'd still need to add this [line](https://github.com/troopa81/docker-images/commit/8552b6f20c4cb53947b2d057762d0fd556079854#diff-142cb0b555bb2f6a7f1eaf4f116931e668735c187d57db3d6423a6b97aa89c4bR87) to make it properly work. 

I try another [approach](https://github.com/troopa81/docker-images/tree/already_configured_complete) which would have not needed to commit and only a call to *buildDockerImage*, but it fails with the following error

```shell
ERROR:
ORA-01017: invalid username/password; logon denied
```

If someone knows why...

**Comment by trancept** [comment]
I use the Docker file from https://github.com/deusaquilus/docker-oracle-xe that initialize the base at build time, then the 'docker run' is fast.
I run it by mounting an existing database:
`docker run -d \
-p 32118:1521 \
-p 35518:5500 \
--name=oracle-xe \
--volume /your-data-folder/oracle-data:/opt/oracle/oradata \
oracle-xe:18c`
May it help some of you.

**Comment by abhisbyk** [comment]
We support [prebuiltdb extension](https://github.com/oracle/docker-images/tree/main/OracleDatabase/SingleInstance/extensions/prebuiltdb) for the same purpose now. You can extend the base image with this extension to create the database and store the datafiles inside the image itself. So, you can run this extended image and it will start in couple of seconds. This extension is supported from 19c onwards.
